### ISSUE TITLE
Updates to ISLCObjectNode's label text

### ISSUE THREAD

**Issue by zepumph** [phetsims/inverse-square-law-common#79]
From work instrumenting GFL for PhET-iO, there are some new problems with the label text for the ISLCObjectNode. Originally the goal was to instrument both the label and shadow text as a single element, but that added too much complexity, and caused https://github.com/phetsims/gravity-force-lab-basics/issues/172. So as a result in https://github.com/phetsims/gravity-force-lab/issues/175 we decided to try to add a stroke around the label, and to get rid of the shadow. Here are the following problems that we are trying to support:

1. Appropriate sizing in all three ISLC sims (aka fixing https://github.com/phetsims/gravity-force-lab-basics/issues/172)
2. Accenting the white text so that it still has good contrast for all mass/charge values in all three sims.
3. A good maxWidth so that in studio, GFL (which has a label like `m1`) can be changed to a longer string without scaling it strangely.

Currently 2 and 3 are in direct opposition to each other. Allow me to explain.

Ideally maxWidth wouldn't be used to clamp the size down of the original string, but instead used to provide a good maximum for i18n (i.e. longer strings). This isn't the case in the current code. maxWidth is setting the size or the english string, such that the PhetFont is irrelevant;no matter the font size, the maxWidth sets the size back down to the size of the english string.  To fix the immediate problem, I increased the maxWidth, and then set the PhetFont accordingly for each sim, generally making them smaller, so that they match the current size of labels. When doing this, we lose the ability to boost the contrast using stroke. This is because the minimum lineWidth for a stroke around the font size needed is too strong. In GFLB PhetFont of 11 works well, but the minimum stroke width looks like this:
![image](https://user-images.githubusercontent.com/6856943/64564143-5f5d0000-d2fd-11e9-9bce-e158dc957d0a.png)

So now it seems like we are stuck between supporting a larger maxWidth for phet-io and i18n, and accenting the text with a stroke. I'll keep investigating.

**Comment by zepumph** [comment]
The issue outlined in the above screenshot was because lineWidth was not supported by RichText. I added it over in https://github.com/phetsims/scenery/issues/998, and that is out for review. Note that there are a few artifacts on the tops of letters, I asked @jonathanolson for his opinion on those over in that issue.

@ariel-phet  @arouinfar  please review how master looks for the three sims, and let me know what you think about the stroke, especially the line width of it. I think that it is a little light for contrast's sake. But any darker looked pretty weird. I would also recommend going into gravity-force-lab studio and changing the mass1Node.labelText.textProperty to see it change, scale, and center.

**Comment by zepumph** [comment]
Also note that the texts should now be back to their original size (re https://github.com/phetsims/gravity-force-lab-basics/issues/172), and confirming that would be good to do also.

**Comment by ariel-phet** [comment]
@zepumph, for lack of a better term, the current text on master looks kind of "janky" and a bit blurry, I don't think it is a reasonable solution yet. 

I suppose I don't know exactly what `TextWithShadow` does, but would it be possible using RichText to just layer a black letter under the white letter, slightly displaced? Is that what TextWithShadow already did?

**Comment by zepumph** [comment]
`TextWithShadow` took two texts, one white and one black, and put them on top of each other, with the black one on bottom and a little offset. The issue was that then we have two texts to instrument and keep in sync. I'm not really sure what the general solution here is. I'm going to mark for developer meeting to see if others have an idea.

For developers, the goal is to have a shadow-like text, but for PhET-iO concerns it is easiest for it to be one text. Is this possible?

**Comment by arouinfar** [comment]
@zepumph thanks for getting the stroked text up and running! I generally agree with @ariel-phet.

When the browser window is maximized, the outline looks pretty good. However, the text is harder to read at smaller resolutions, and doesn't scale very for long strings (even in a maximized window). 

The shadow text is easier to read when the text is small.
![image](https://user-images.githubusercontent.com/8419308/64730436-94538900-d49c-11e9-9d31-b37203255b38.png)
![image](https://user-images.githubusercontent.com/8419308/64730444-99b0d380-d49c-11e9-9483-95c8b8caffad.png)

**Comment by zepumph** [comment]
@pixelzoom had a helpful comment over in https://github.com/phetsims/scenery/issues/998#issuecomment-530568890:

>Imo, stroking a font (in general) looks lousy. You'd be better off filling bold text with a color that looks good on each sphere, or putting a translucent background behind filled text.

**Comment by arouinfar** [comment]
>You'd be better off filling bold text with a color that looks good on each sphere

In principle, I agree, but that's not possible here. With the Constant Size option, the color varies from nearly white to fully saturated red or blue. Black text would look best for low densities, but becomes unreadable when the color of the mass is fully saturated. White text would look best for high densities, but would be impossible to read at the lower end of the spectrum.

> translucent background behind filled text

I'd be fine with giving this a try to see how it looks. I think black text with a transparent white background would be ideal.

**Comment by samreid** [comment]
We should continue the development of SCENERY_PHET/ShadowText.js and instrument it (without instrumenting the inner 2 Text instances--they are an implementation detail).  We should strive to give our Text types similar PhET-iO interfaces: RichText, ShadowText, Text (maybe others?) should have `setText`, `setFontOptions`, `setMaxWidth`, etc.

**Comment by arouinfar** [comment]
@zepumph I'm liking the transparent background! The padding, corner radius, and bold font all look good to me. I would like to bump the opacity up to 0.6, to help boost the contrast of the text when the color is fully saturated.

As an aside, I don't love how the labels spill over a bit in constant size mode in GFLB. Could we bump up the size of the masses just a tiny bit? If the size in GFL and GFLB are the same, the change could apply to both sims. @zepumph if you'd like to move this request to a separate issue, please let me know.

![image](https://user-images.githubusercontent.com/8419308/64890727-884b0100-d62d-11e9-9f03-f84c86ad549a.png)

**Comment by zepumph** [comment]
In the above commits I did the following work:

* Added a common code type called BackgroundNode, which is used by ISLC
* made the GFLB constant radius value bigger (not GFL).
* Tweaked Coulombs Law to try to make that look alright under the current conditions discussed above.
  * One of these tweaks was to place the labels on the atomic scale at a bit of an offset, this way they didn't overlap. This was easier than detecting the overlap, but if this is not acceptable I understand, and let's create a new issue to handle that (only if you think it is needed). Here is what it currently looks like:
  ![image](https://user-images.githubusercontent.com/6856943/64897245-3905d000-d62f-11e9-8dcb-40cff35bb69b.png)
![image](https://user-images.githubusercontent.com/6856943/64897306-6eaab900-d62f-11e9-8224-509af894817e.png)


Please review all three sims. Does this seem good to you?

**Comment by ariel-phet** [comment]
@zepumph @arouinfar I generally like the direction, however, a few tweaks

1. @zepumph on Coulomb's Law on the Atomic scale, things don't look very good. I think for that screen just go with white lettering and no stroke (the visual appearance of the published sim), That seems far superior

2. On Coulomb's Law on the Macro Scale I think darkening the background slightly would be nice for the text when you are on the case of zero charge, but it probably works as is.

3. I agree with @arouinfar that on GFL:B we should either bump up the minimum mass size slightly or reduce the text label size slightly. Increasing the mimimum mass size seems like the better option.

4. For GFL seems the background needs to be a bit more gray again for contrast for small masses

**Comment by zepumph** [comment]
(1) should be completed in the above commit.

**Comment by zepumph** [comment]
>I agree with @arouinfar that on GFL:B we should either bump up the minimum mass size slightly or reduce the text label size slightly. Increasing the mimimum mass size seems like the better option.

During my prototyping, I accidentally committed a change making the font bold. I removed that above, and I think that fixed the sizing discrepancies for GFL:B.

(2) and (4) both seemed like they could be solved by increasing the opacity, which I did generally (simplifying the ISLC code to use the same default as the common code type). Now all the backgrounds should be a bit darker. The same as for backgrounds of the equations in Graphing Quadratics.

@ariel-phet how to the three sims look now?

**Comment by arouinfar** [comment]
> (1) should be completed in the above commit.

Coulomb's Law is looking nice!

> During my prototyping, I accidentally committed a change making the font bold. I removed that above, and I think that fixed the sizing discrepancies for GFL:B.

It's definitely an improvement, but I think I would like to see the mass in Constant Size mode ever-so-slightly bigger so there's a bit of negative space between the corner of the label and the edge of the mass. This is still looking pretty tight to me.

![image](https://user-images.githubusercontent.com/8419308/65085740-77bbc300-d96c-11e9-86f1-78c24fb85afd.png)


> (2) and (4) both seemed like they could be solved by increasing the opacity, which I did generally (simplifying the ISLC code to use the same default as the common code type). Now all the backgrounds should be a bit darker. The same as for backgrounds of the equations in Graphing Quadratics.

The more opaque background looks fine to me, but it looks a bit too opaque to me. I don't have a strong feeling about this though, so I'm happy to let @ariel-phet make that call.

**Comment by ariel-phet** [comment]
@arouinfar @zepumph I think I am just not liking the way these backgrounds look when the mass gets slightly smaller than the background text box. It also seems like the moment someone changes that text in iO, it just won't work well.

(Coulomb's law looks fine currently)

What about if we just move the label off the masses, and just have it "stick" a fixed distance from the bottom center of the masses? It seems like that will allow for a longer strings, plenty of flexibility (and slightly more complicated code...)

I tried a little mockup, and I think it looks fine:

![GFL](https://user-images.githubusercontent.com/4051174/65162907-4b995400-d9f7-11e9-97a0-c5bb297229e9.PNG)

It just seems like that would be a more general solution, and less fiddly in the long run.

**Comment by ariel-phet** [comment]
And obviously above, the font sizes could probably be a bit smaller, and things could be shifted a bit if we wanted more negative space for the ruler.

**Comment by arouinfar** [comment]
> What about if we just move the label off the masses, and just have it "stick" a fixed distance from the bottom center of the masses? It seems like that will allow for a longer strings, plenty of flexibility (and slightly more complicated code...)

@ariel-phet @zepumph I like that idea! I would keep the text the same, but relocate it to the bottom of the mass. The label positioning in @ariel-phet's mockup looks good. 

Another way we can optimize for longer strings, would be to have smart labels that shift away from one another when the masses are close. This certainly isn't critical, but would be a nice touch.

I could see a client wanting to change the labels on the charges to something like `charge 1` and `charge 2`, which will become unreadable when the masses are close. 
![image](https://user-images.githubusercontent.com/8419308/65163716-aaab9880-d9f8-11e9-9679-e4b948829c4d.png)

**Comment by arouinfar** [comment]
@zepumph I saw this sim went up for rc in https://github.com/phetsims/QA/issues/429 (exciting!), but I think we'll want to include the label positions suggested by @ariel-phet in https://github.com/phetsims/inverse-square-law-common/issues/79#issuecomment-532740185 in the 1.0 release. 

The label shifting to avoid overlaps isn't necessary for 1.0, and if we decide to go there, could get folded into the PhET-iO work.

**Comment by zepumph** [comment]
Agreed! I was expecting to patch in the final touches to this issue, but I wanted to get that test going sooner.

**Comment by zepumph** [comment]
I added the labels below the object in the above commit. This involved small tweaks to the default ruler positions. It is possible, like @arouinfar stated above, to have labels overlap under certain conditions with longer labels specified by PhET-iO. It doesn't bother me too much though.

@arouinfar @ariel-phet how does this look?

**Comment by ariel-phet** [comment]
@zepumph looking pretty good to me. Just one minor tweak:

- In Coulomb's Law, Macro Scale screen, I would center the ruler vertically in the negative space (right now, default position is a little too close to the charge labels)

**Comment by arouinfar** [comment]
The labels look great @zepumph.

> In Coulomb's Law, Macro Scale screen, I would center the ruler vertically in the negative space (right now, default position is a little too close to the charge labels)

Agreed, I would shift the ruler down to about here:

![image](https://user-images.githubusercontent.com/8419308/65447300-74ed2200-ddf3-11e9-8dec-b9eb070eafa1.png)

**Comment by zepumph** [comment]
Thanks. Implemented above. I'll cherrypick what's needed into the rc now. Closing
